packages/fetchai/skills/carpark_client/behaviours.py

```diff
@@ -19,11 +19,99 @@
 
 """This package contains the behaviours of the agent."""
 
+from typing import Any, cast
+
+from aea.skills.behaviours import TickerBehaviour
+
+from packages.fetchai.protocols.fipa.message import FipaMessage
+from packages.fetchai.skills.carpark_client.strategy import Strategy
 from packages.fetchai.skills.generic_buyer.behaviours import (
     GenericSearchBehaviour,
     GenericTransactionBehaviour,
 )
+from packages.fetchai.skills.generic_buyer.dialogues import FipaDialogue, FipaDialogues
 
 
 SearchBehaviour = GenericSearchBehaviour
 TransactionBehaviour = GenericTransactionBehaviour
+
+
+class ProposalCheckBehaviour(TickerBehaviour):
+    """
+    Controls the timer to check if proposals have been received.
+
+    :param TickerBehaviour: generic ticker behaviour to be inherited from
+    """
+
+    def __init__(self, **kwargs: Any) -> None:
+        """Initialize the proposal check behaviour."""
+        super().__init__(**kwargs)
+        self.counter = 0
+
+    def setup(self) -> None:
+        """Set up the behaviour."""
+
+    def act(self) -> None:
+        """Act according to the behaviour."""
+        strategy = cast(Strategy, self.context.strategy)
+        if not strategy.waiting_for_proposals:
+            return
+
+        # check if every agent has sent at least one proposal
+        if set(strategy.sent_proposals).issubset(
+            list(map(lambda x: x["sender"], strategy.received_proposals))
+        ):
+            self.context.logger.info("received all proposals, making decision...")
+        elif self.counter > strategy.proposal_check_timeout - 1:
+            self.context.logger.info(
+                "waiting for proposals timed out, making decision..."
+            )
+        else:
+            self.counter += 1
+            return
+        self._handle_all_proposals_received(strategy)
+
+    def teardown(self) -> None:
+        """Teardown the behaviour."""
+
+    def _handle_all_proposals_received(self, strategy: Strategy) -> None:
+        undecided_proposals = list(
+            filter(lambda x: x["decision"] is None, strategy.received_proposals)
+        )
+        """Prepare the accept/decline messages"""
+        if undecided_proposals:
+            cheapest_proposal = strategy.get_cheapest_proposal(undecided_proposals)
+            for carpark in undecided_proposals:
+                cheapest = cheapest_proposal["sender"] == carpark["sender"]
```

Comment:
Some of the variables could have better names, such as `carpark["sender"]`.

Reply:
You are right, I was planning to rename the variables to be more generic and for them to ultimately end up in the generic buyer logic. Those fell under the radar and they will be changed.